README.rst
======
lwotai
======
LWOTai is an AI program that provides the opponent in the board game "Labyrinth: the War on Terror, 2001 - ?".


Description
===========

LWOTai is a command line program. You need to have Python installed on your system to use it. You can run it from a command line as follows: python lwotai.py The most important thing to know is that the program does not track the deck, cards, hands or use of Reserves. You have to do that either with the board game or on Vassal.
Set up the game as usual and pick the scenario and ideology. You will then be at the Command prompt. You can enter "?" to see the various commands. The most important are the "j" and "u" commands. When it is a player's turn, pick the card they will play and enter that command and the card number. For instance, it is the Jihadist turn at the beginning of the game. Turn over the top card of their hand and then enter j and that card number (e.g. j 22). The program will then tell you what the Jihadist did so you can update your map. It's the same with your plays using the u command. If more information is necessary the program will ask for it.

When you use a card for Ops you then use the command for the operation you want. For instance if you want to disrupt, you enter "disrupt" at the Command prompt. The program will then ask what it needs to know about the disrupt.

There are two timekeeping functions you have to remember to use. After each US Activity Phase, you must enter the "plot" command so that unblocked plots are handled. And at the end of a turn you must enter the "turn" command to handle all the end of turn activities.

At any time you can use the "status" command to get a printout of the entire board position. And you can use the "history" command to see everything that has happened in the game.

Thanks to Dave Horn for implementing the Save and Undo system.

1. A save game is created after every single command whether you want it or not. If someone screws up and closes the window, PC battery dies, crashes, whatever, no problem, load it up again and you will be asked if you want to load the suspended game.

2. Rollback files are created at the beginning of each turn. You can roll back to any previous turn using the 'roll' or 'rollback' command. You will be prompted to enter which turn you want to roll back to.

3. An undo file is created after every card played. The player can undo to the last card at any time (two exceptions) by typing 'undo'. Exceptions are when you load from a previously suspended game or after executing a rollback. The undo file is removed at that exact point to prevent the player from undoing themselves to some other game in the past!

Thanks to Peter Shaw for implementing the Adjust system and for a bunch of bug fixes and cleanup.




CHANGES.rst
=========
Changelog
=========

Version 1.11082014.1 (by sgwestrip)
===================================

Changed wording for Jihadist Ideology choices. The wording is now as per the rulebook.
Added: release number displays when you first run the program.

Version 1.06082014.2 (by ragam)
===============================

Fix: cards #84/85 will now make you Shift Ally to Neutral, if possible.
Fix: card #58 will now correctly Disrupt cadres if the event is in play

Version 1.04082014.1 (by sgwestrip)
===================================

Add: new Jihadist Ideology (9.7) Coherent. This was added to the rules after the initial printing.
Fix: card #75 now displays 'Remove card from game' when it is played for the event.
Fix: numerous typos
Fix: remove all "Jihadist Activity Phase finished, enter plot command." as this is incorrect. It should always be done after the 2nd US Activity Phase.

Version 1.06162015.1
====================
1. Made use of CTR consistent - was adding CRT but testing for CTR (reported by Morten Kristensen)

Version 1.20150312.1
====================
1. Fixed missing line for card 102 (reported by Thomas Chipman)

Version 1.20150303.1
====================
1. Prevent double processing when major jihad failure sets besieged status (reported by Magnus Kvevlander)

Version 1.20150131.1
====================
1. Fixed spelling of besieged in Country class
2. Added untested countries with data to 'status' command so that 'status' can be used to reconstruct the board
3. Fixed 'help dep'
4. Added valid global markers, country markers and lapsing markers for use by 'adjust' command
5. Added 'adjust' command for ideology, prestige, funding, event markers, lapsing markers and country data
6. Additional changes for multiple aid markers when adding or removing
7. In plot resolution, remove one aid for each successful roll
8. In WoI roll adjustment, add 1 to modifiers for each aid marker
9. Ignore "The door of Itjihad is closed" when checking playable if playing as US (reported by Dave Horn)
10. Changed removeCell to remove sleeper then Sadr then active if playing as US; but active then sleeper then Sadr if Jihadist
11. Fixed test for infectious ideology when setting difficulty
12. Added 'summary' command to display state of each track
13. Changed message if no Islamist Rule countries at 'turn'
14. Changed message when card 1 event activated

See GitHub for changes submitted via pull request, e.g. by andrewswan.


lwotai/deck.py
from lwotai.card import Card


class Deck(object):
    """The deck of cards in the game"""

    def __init__(self, cards):
        self.cards = {}
        # Store the given list of cards in a dict, indexed by their number
        for card in cards:
            self.cards[card.number] = card

    def get(self, card_number):
        """Returns the card with the given number"""
        assert isinstance(card_number, int)
        card = self.cards[card_number]
        if not card:
            raise ValueError("Invalid card number %d" % card_number)
        return card


CARDS = [
    Card(1, "US", "Backlash", 1, False, False, False),
    Card(2, "US", "Biometrics", 1, False, False, True),
    Card(3, "US", "CTR", 1, False, True, False),
    Card(4, "US", "Moro Talks", 1, True, True, False),
    Card(5, "US", "NEST", 1, True, True, False),
    Card(6, "US", "Sanctions", 1, False, False, False),
    Card(7, "US", "Sanctions", 1, False, False, False),
    Card(8, "US", "Special Forces", 1, False, False, False),
    Card(9, "US", "Special Forces", 1, False, False, False),
    Card(10, "US", "Special Forces", 1, False, False, False),
    Card(11, "US", "Abbas", 2, True, True, False),
    Card(12, "US", "Al-Azhar", 2, False, False, False),
    Card(13, "US", "Anbar Awakening", 2, False, True, False),
    Card(14, "US", "Covert Action", 2, False, False, False),
    Card(15, "US", "Ethiopia Strikes", 2, True, False, False),
    Card(16, "US", "Euro-Islam", 2, True, False, False),
    Card(17, "US", "FSB", 2, False, False, False),
    Card(18, "US", "Intel Community", 2, False, False, False),
    Card(19, "US", "Kemalist Republic", 2, False, False, False),
    Card(20, "US", "King Abdullah", 2, True, False, False),
    Card(21, "US", "Let's Roll", 2, False, False, False),
    Card(22, "US", "Mossad and Shin Bet", 2, False, False, False),
    Card(23, "US", "Predator", 2, False, False, False),
    Card(24, "US", "Predator", 2, False, False, False),
    Card(25, "US", "Predator", 2, False, False, False),
    Card(26, "US", "Quartet", 2, False, False, False),
    Card(27, "US", "Saddam Captured", 2, True, True, False),
    Card(28, "US", "Sharia", 2, False, False, False),
    Card(29, "US", "Tony Blair", 2, True, False, False),
    Card(30, "US", "UN Nation Building", 2, False, False, False),
    Card(31, "US", "Wiretapping", 2, False, True, False),
    Card(32, "US", "Back Channel", 3, False, False, False),
    Card(33, "US", "Benazir Bhutto", 3, True, True, False),
    Card(34, "US", "Enhanced Measures", 3, False, True, False),
    Card(35, "US", "Hijab", 3, True, False, False),
    Card(36, "US", "Indo-Pakistani Talks", 3, True, True, False),
    Card(37, "US", "Iraqi WMD", 3, True, True, False),
    Card(38, "US", "Libyan Deal", 3, True, True, False),
    Card(39, "US", "Libyan WMD", 3, True, True, False),
    Card(40, "US", "Mass Turnout", 3, False, False, False),
    Card(41, "US", "NATO", 3, False, True, False),
    Card(42, "US", "Pakistani Offensive", 3, False, False, False),
    Card(43, "US", "Patriot Act", 3, True, True, False),
    Card(44, "US", "Renditions", 3, False, True, False),
    Card(45, "US", "Safer Now", 3, False, False, False),
    Card(46, "US", "Sistani", 3, False, False, False),
    Card(47, "US", "The door of Itjihad was closed", 3, False, False, True),
    Card(48, "Jihadist", "Adam Gadahn", 1, False, False, False),
    Card(49, "Jihadist", "Al-Ittihad al-Islami", 1, True, False, False),
    Card(50, "Jihadist", "Ansar al-Islam", 1, True, False, False),
    Card(51, "Jihadist", "FREs", 1, False, False, False),
    Card(52, "Jihadist", "IEDs", 1, False, False, False),
    Card(53, "Jihadist", "Madrassas", 1, False, False, False),
    Card(54, "Jihadist", "Moqtada al-Sadr", 1, True, True, False),
    Card(55, "Jihadist", "Uyghur Jihad", 1, True, False, False),
    Card(56, "Jihadist", "Vieira de Mello Slain", 1, True, True, False),
    Card(57, "Jihadist", "Abu Sayyaf", 2, True, True, False),
    Card(58, "Jihadist", "Al-Anbar", 2, True, True, False),
    Card(59, "Jihadist", "Amerithrax", 2, False, False, False),
    Card(60, "Jihadist", "Bhutto Shot", 2, True, True, False),
    Card(61, "Jihadist", "Detainee Release", 2, False, False, False),
    Card(62, "Jihadist", "Ex-KGB", 2, False, False, False),
    Card(63, "Jihadist", "Gaza War", 2, False, False, False),
    Card(64, "Jihadist", "Hariri Killed", 2, True, False, False),
    Card(65, "Jihadist", "HEU", 2, True, False, False),
    Card(66, "Jihadist", "Homegrown", 2, False, False, False),
    Card(67, "Jihadist", "Islamic Jihad Union", 2, True, False, False),
    Card(68, "Jihadist", "Jemaah Islamiya", 2, False, False, False),
    Card(69, "Jihadist", "Kazakh Strain", 2, True, False, False),
    Card(70, "Jihadist", "Lashkar-e-Tayyiba", 2, False, False, False),
    Card(71, "Jihadist", "Loose Nuke", 2, True, False, False),
    Card(72, "Jihadist", "Opium", 2, False, False, False),
    Card(73, "Jihadist", "Pirates", 2, True, True, False),
    Card(74, "Jihadist", "Schengen Visas", 2, False, False, False),
    Card(75, "Jihadist", "Schroeder & Chirac", 2, True, False, False),
    Card(76, "Jihadist", "Abu Ghurayb", 3, True, False, False),
    Card(77, "Jihadist", "Al Jazeera", 3, False, False, False),
    Card(78, "Jihadist", "Axis of Evil", 3, False, False, False),
    Card(79, "Jihadist", "Clean Operatives", 3, False, False, False),
    Card(80, "Jihadist", "FATA", 3, False, True, False),
    Card(81, "Jihadist", "Foreign Fighters", 3, False, False, False),
    Card(82, "Jihadist", "Jihadist Videos", 3, False, False, False),
    Card(83, "Jihadist", "Kashmir", 3, False, False, False),
    Card(84, "Jihadist", "Leak", 3, False, False, False),
    Card(85, "Jihadist", "Leak", 3, False, False, False),
    Card(86, "Jihadist", "Lebanon War", 3, False, False, False),
    Card(87, "Jihadist", "Martyrdom Operation", 3, False, False, False),
    Card(88, "Jihadist", "Martyrdom Operation", 3, False, False, False),
    Card(89, "Jihadist", "Martyrdom Operation", 3, False, False, False),
    Card(90, "Jihadist", "Quagmire", 3, False, False, False),
    Card(91, "Jihadist", "Regional al-Qaeda", 3, False, False, False),
    Card(92, "Jihadist", "Saddam", 3, False, False, False),
    Card(93, "Jihadist", "Taliban", 3, False, False, False),
    Card(94, "Jihadist", "The door of Itjihad was closed", 3, False, False, False),
    Card(95, "Jihadist", "Wahhabism", 3, False, False, False),
    Card(96, "Unassociated", "Danish Cartoons", 1, True, False, False),
    Card(97, "Unassociated", "Fatwa", 1, False, False, False),
    Card(98, "Unassociated", "Gaza Withdrawal", 1, True, False, False),
    Card(99, "Unassociated", "HAMAS Elected", 1, True, False, False),
    Card(100, "Unassociated", "Hizb Ut-Tahrir", 1, False, False, False),
    Card(101, "Unassociated", "Kosovo", 1, False, False, False),
    Card(102, "Unassociated", "Former Soviet Union", 2, False, False, False),
    Card(103, "Unassociated", "Hizballah", 2, False, False, False),
    Card(104, "Unassociated", "Iran", 2, False, False, False),
    Card(105, "Unassociated", "Iran", 2, False, False, False),
    Card(106, "Unassociated", "Jaysh al-Mahdi", 2, False, False, False),
    Card(107, "Unassociated", "Kurdistan", 2, False, False, False),
    Card(108, "Unassociated", "Musharraf", 2, False, False, False),
    Card(109, "Unassociated", "Tora Bora", 2, True, False, False),
    Card(110, "Unassociated", "Zarqawi", 2, False, False, False),
    Card(111, "Unassociated", "Zawahiri", 2, False, False, False),
    Card(112, "Unassociated", "Bin Ladin", 3, False, False, False),
    Card(113, "Unassociated", "Darfur", 3, False, False, False),
    Card(114, "Unassociated", "GTMO", 3, False, False, True),
    Card(115, "Unassociated", "Hambali", 3, False, False, False),
    Card(116, "Unassociated", "KSM", 3, False, False, False),
    Card(117, "Unassociated", "Oil Price Spike", 3, False, False, True),
    Card(118, "Unassociated", "Oil Price Spike", 3, False, False, True),
    Card(119, "Unassociated", "Saleh", 3, False, False, False),
    Card(120, "Unassociated", "US Election", 3, False, False, False)
]
